### ISSUE TITLE
Spring Boot WebMVC Autoconfigure support

### ISSUE THREAD

**Issue by agentgt** [jstachio/jstachio#169]
Placeholder for now for continued work on what was originally #10 .

May depend on #168 . Support for this may happen before 2.0.0 so ignore milestone for now.

**Comment by dsyer** [comment]
See #173 for some code ideas.

**Comment by agentgt** [comment]
@dsyer I'm thinking we could add some support for String based views and an early version of context aware templates for boot for those that still prefer `Map` based models.

Old school MVC works like:

```java
@GetMapping("/")
public String controller(Map<String,Object> model) {
  SomeJStacheModel m = new SomeJStacheModel(...);
  return JStachioModelView.of(model, m).asStringView();
}
```

I guess this is less about Spring Boot and more about helping existing legacy Mustache applications (JMustache or mustache.java) port over.

Then again maybe it is not worth it.

**Comment by dsyer** [comment]
I'm not sure I see the benefit of a String-based view that has to be derived from something that already knows how to render itself. Maybe if you can return a String literal, which we could still support I guess, but I don't see how to do that without thread locals or request scope. Maybe you had an implementation in mind? A convention-based `Model` entry? I don't like that much, but I've seen it in other "reflection-free" template engines.

The context-aware templates definitely seem like a good idea.

**Comment by damir78** [comment]
I would suggest the simple idea: write autoconfiguration based on official existing spring-boot-starter-mustache: 
1) https://github.com/spring-projects/spring-boot/blob/v3.1.0/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mustache/MustacheAutoConfiguration.java
2) https://github.com/spring-projects/spring-boot/blob/v3.1.0/spring-boot-project/spring-boot-starters/spring-boot-starter-mustache/build.gradle

**Comment by dsyer** [comment]
Did you notice the author tags on that file?

**Comment by agentgt** [comment]
I actually already took ideas from it like making a pluggable TemplateFinder (albeit in JMustache it is called TemplateLoader).

I just need to get rid of SpringTemplateConfig and merge it with main autoconfig.


(And yes I noticed the author :smile: )

**Comment by Allsimon** [comment]
Is hot reload with spring-boot-devtools in the scope of this issue ?

**Comment by dsyer** [comment]
Devtools is not really about autoconfiguration (much), so arguably not. Did you try it and it doesn't work?

**Comment by agentgt** [comment]
Hot reload in general is tricky business. 

That is the number one focus for v1.3 .

We have fake hot reload with jmustache. If you add io.jstach.opt.jmustache aka jstachio-jmustache to the classpath it will try to use jmustache to render changed files.

But this only works for changing template files and not models or online templates. JMustache is also not equal to what JStachio knows of mustache syntax.


Im still researching solutions.

**Comment by dsyer** [comment]
I don't like the JMustache fake thing so I'm glad you're thinking about it. For me it works if you remember to change the template before the model (I suppose it would be better if a change to the template was sufficient to trigger a re-compilation). We should open another issue though because it really isn't to do with autoconfig.

**Comment by agentgt** [comment]
Spring Boot Webflux will be 1.3.

**Comment by agentgt** [comment]
@dsyer I have re-deploy 1.2.0 as a SNAPSHOT in central which includes Spring Boot work. 

When you get a chance check it out and let me know what you think.

**Comment by dsyer** [comment]
I tried the snapshots last week. Did something change?

**Comment by agentgt** [comment]
Yes. Mainly just performance on output and additional configuration on buffer limit.

I tried fixing up TechEmpower Spring webflux to test JStachio but it is a mess and for some reason I can't get the `tfb` script to execute it spring-webflux.

I will try TechEmpowers Spring (mvc) benchmark later.

**Comment by dsyer** [comment]
I'm only using MVC but it works same as before. There's no autoconfig for Webflux yet, right?

**Comment by agentgt** [comment]
> I'm only using MVC but it works same as before. There's no autoconfig for Webflux yet, right?

That is correct. Do you want that now for 1.2 which I will release next week or can we wait?

I have not re-published a snapshot for this but I moved the static config to shard import class and moved the mustache template files to `src/main/resources/templates` like it is for for FTL and Mustache in Spring Boot.

I made https://jstach.io/jstachio/io.jstach.opt.spring/io/jstach/opt/spring/SpringJStacheConfig.html

Which is a shared Spring Boot like static config for JStachio. Remember you helped me come up with the brilliant idea of config sharing with `using`. I'm so thankful you convinced me that was a problem that needed fixing.

So now ignoring Maven setup the hell that is module-info a Spring Boot user can just add the starter as a dep.

Then they just put 

```java
 // package-info.java
 @JStacheConfig(using = SpringJStacheConfig.class)
 package some.package;
```

On any package that has JStache models.

It is annoying they have to put it on each package they have models in but hopefully one day module-info will be more embraced (probably the same time people stop using Java 8 aka never).

**Comment by dsyer** [comment]
I think it can wait. I was just asking. Sounds like things are moving in a good direction.
